### PR TITLE
feat: use internal cert

### PR DESCRIPTION
**Pull request by charles-chenzz**
solve #52 

might missing something or did anything wrong, if so please correct me.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -88,16 +89,46 @@ func main() {
...
+
+	setupLog.Info("starting manager")
+	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
+		setupLog.Error(err, "problem running manager")
+		os.Exit(1)
+	}
+}
+
+func setupControllers(mgr ctrl.Manager, certsReady chan struct{}) {
+	// wait until the cert ready, until then we set up controllers
```
Comment (danielvegamyhre) [inline main.go:113]:
```suggestion
	// The controllers won't work until the webhooks are operating, and the webhook won't work until the
	// certs are all in place.
```

**Comment by danielvegamyhre** [comment]
@charles-chenzz looks good, a couple things:

1. You need to update the Dockerfile to copy `/pkg/util/cert` into the image as well.
2. Please update the installation instructions in the README to remove the step requiring the user to install `cert-manager`

File: main.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ package main
 import (
 	"flag"
 	"os"
+	"sigs.k8s.io/jobset/pkg/util/cert"
```
Comment (ahg-g) [inline main.go:22]: move this to the middle block of imports below

File: test/integration/webhook/suite_test.go [hunk 3]
```diff
@@ -35,8 +35,8 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/envtest"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
-
```
Comment (ahg-g) [inline test/integration/webhook/suite_test.go:38]: keep this empty line, it separates our imports from the rest

File: test/integration/webhook/suite_test.go [hunk 4]
```diff
@@ -35,8 +35,8 @@ import (
...
 	jobset "sigs.k8s.io/jobset/api/v1alpha1"
+	"sigs.k8s.io/jobset/pkg/controllers"
```
Comment (ahg-g) [inline test/integration/webhook/suite_test.go:39]: put this in the block of imports above

File: main.go [hunk 5]
```diff
@@ -107,10 +138,10 @@ func main() {
 		setupLog.Error(err, "unable to set up ready check")
 		os.Exit(1)
 	}
+}
 
-	setupLog.Info("starting manager")
-	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
-		setupLog.Error(err, "problem running manager")
-		os.Exit(1)
+func setupIndexes(mgr ctrl.Manager) {
```
Comment (ahg-g) [inline main.go:143]: no need for this function, it is only a one liner.

File: main.go [hunk 6]
```diff
@@ -88,16 +89,46 @@ func main() {
...
+
+	certsReady := make(chan struct{})
+	if err = cert.CertsManager(mgr, certsReady); err != nil {
+		setupLog.Error(err, "unable to setup cert rotation")
+		os.Exit(1)
+	}
+
+	setupIndexes(mgr)
+
+	go setupControllers(mgr, certsReady)
```
Comment (ahg-g) [inline main.go:101]: why do we have to setup the controllers in a separate go routine?

**Review by ahg-g — COMMENTED** [review]
Thank you very much, looking forward for more contributions!

Did you test this manually by any chance?

File: main.go (hunk 1 above)
Comment (charles-chenzz) [reply main.go]: done

File: main.go (hunk 2 above)
Comment (charles-chenzz) [reply main.go]: done

File: test/integration/webhook/suite_test.go (hunk 3 above)
Comment (charles-chenzz) [reply test/integration/webhook/suite_test.go:38]: done

File: test/integration/webhook/suite_test.go (hunk 4 above)
Comment (charles-chenzz) [reply test/integration/webhook/suite_test.go:40]: done

File: main.go (hunk 5 above)
Comment (charles-chenzz) [reply main.go]: done

File: main.go (hunk 6 above)
Comment (charles-chenzz) [reply main.go:103]: controllers are blocked by certReady chan

File: pkg/util/cert/cert.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cert
+
+import (
+	"fmt"
```
Comment (ahg-g) [inline pkg/util/cert/cert.go:17]: pls add new line below to split the builtin pkgs from the rest

File: main.go (hunk 6 above)
Comment (ahg-g) [reply main.go:103]: I get that part. 

I guess the part I missed is that the CertsManager call above registers a reconciler with the manager that gets started only when we invoke `mgr.Start` below

**Review by ahg-g — COMMENTED** [review]
one nit

/lgtm

leaving approve to Daniel

File: pkg/util/cert/cert.go (hunk 7 above)
Comment (charles-chenzz) [reply pkg/util/cert/cert.go:17]: done

File: main.go (hunk 6 above)
Comment (ahg-g) [reply main.go:103]: +1, can you please add the commend Daniel suggested above?

Comment (charles-chenzz) [reply main.go:106]: done. please check again

Comment (danielvegamyhre) [reply main.go:106]: thanks!

**Pull request merged**